enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,502 @@
+order to simplify a transition. 
+
+The agent uses an OCI Hook to enable swap via container cgroup. The hook script 
+sets the containers cgroup `memory.swap.max=max`.
+
+* **Tech Preview**
+  * Limited to burstable QoS class pods.
+  * Uses `UnlimitedSwap` approach. Using OCI hook each container is started with `memory.swap.max=max` in its cgroup.
+* **General Availability**
+  * Container starts with `UnlimitedSwap` as in TP, then reconciled to `LimitedSwap` by the wasp-agent limited swap controller.
+    * For more info, refer to the upstream documentation on how to calculate [limited swap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap#steps-to-calculate-swap-limit). 
+  * Added Limitation to non-high-priority pods.
+  * Added swap-based evictions.
+  * Added critical workload protection (same as kube swap).
+
+![wasp](https://github.com/user-attachments/assets/314d387d-788a-4aa6-9275-c882f5f329aa)
+
+**NOTE**: In WASP GA setting the swap limit isn't atomic, it can happen that a container application will start
+to allocate memory before the limit is set by the wasp-agent.
+
+###### Machine-level swap configuration
+
+* Provisioning swap
+
+Provisioning of swap is left to the cluster administrator.
+The OCI hook itself is not making any assumption where the swap is located.
+
+As long as there is no additional tooling available, the recommendation
+is to use `MachineConfig` objects to provision swap on nodes.
+
+The `MachineConfig` object would include the necessary scripts
+in order to provision swap on a disk, partition, or in a file.
+
+* Node service protection
+
+All container workloads are run in the `kubepods.slice` cgroup.
+All system services are run in the `system.slice` cgroup.
+
+By default the `system.slice` is permitted to swap, however, system
+services are critical for a node's health and in turn critical for the
+workloads.
+
+Without system services such as `kubelet` or `crio`, any container will
+not be able to run well.
+
+Thus, in order to protect the `system.slice` and ensure that the nodes
+infrastructure health is prioritized over workload health, Machine Config is
+used to configure `memory.swap.max` to zero on the system slice parent cgroup.
+The configuration is done via systemd for the sake of consistency.
+
+* Preventing SWAP traffic I/O saturation
+
+One risk of heavy swapping is to saturate the disk bus with swap traffic,
+potentially preventing other processes from performing I/O.
+
+In order to ensure that system services are able to perform I/O, the
+machine config is configuring `io.latency=50` for the `system.slice` in order
+to ensure that its I/O requests are prioritized over any other slice.
+This is, because by default, no other slice is configured to have
+`io.latency` set.
+
+* Node memory pressure handling
+
+Dealing with memory pressure on a node is differentiating the TP fom GA.
+
+** **Technology Preview** - `memory.high` is set on the `kubepods.slice`
```

Comment:
has this been implemented?

Reply:
It was. @enp0s3 @iholder101 is it still?